src/v/cluster/self_test/netcheck.cc

```diff
@@ -0,0 +1,166 @@
+  : _self(self)
+  , _connections(connections) {}
+
+ss::future<> netcheck::start() { return ss::now(); }
+
+ss::future<> netcheck::stop() {
+    _as.request_abort();
+    return _gate.close();
+}
+
+void netcheck::cancel() { _cancelled = true; }
+
+ss::future<std::vector<self_test_result>> netcheck::run(netcheck_opts opts) {
+    vassert(_gate.get_count() == 0, "Benchmark already in progress");
+    gate_guard g{_gate};
+    _cancelled = false;
+    _opts = opts;
+    try {
+        co_return co_await ss::with_scheduling_group(opts.sg, [this]() {
+            return ssx::async_transform(
+              _opts.peers, [this](model::node_id peer) {
+                  return run_individual_benchmark(peer);
+              });
+        });
+    } catch (const network_benchmark_exception& ex) {
+        vlog(clusterlog.error, "Network exception encountered: {}", ex);
+        throw;
+    } catch (const std::exception& ex) {
+        vlog(clusterlog.error, "Unforseen exception encountered: {}", ex);
+        throw network_benchmark_exception(ex.what());
+    }
+}
+
+ss::future<self_test_result>
+netcheck::run_individual_benchmark(model::node_id peer) {
+    auto irange = boost::irange<uint16_t>(0, _opts.parallelism);
+    static const auto two_seconds_us = 200000;
+    const auto now = ss::lowres_clock::now();
+    metrics m{two_seconds_us};
+    vassert(
+      _opts.max_duration >= _opts.duration, "Misconfigured self_test plan");
+    const auto max_timeout = now + _opts.max_duration;
+    co_await ss::parallel_for_each(irange, [this, max_timeout, peer, &m](auto) {
+        return run_benchmark_fiber(
+          run_fiber_opts(max_timeout, _opts.duration), peer, m);
+    });
+    m.set_total_time(ss::lowres_clock::now() - now);
+    auto result = m.to_st_result();
+    result.name = _opts.name;
+    result.info = fmt::format("Test performed against node: {}", peer);
+    if (_cancelled) {
+        result.warning = "Run was manually canclled";
+    }
+    co_return result;
+}
+
+ss::future<> netcheck::run_benchmark_fiber(
+  run_fiber_opts fiber_state, model::node_id peer, metrics& m) {
+    /// run_fiber_opts manages the the RPC request timeouts to ensure that
+    /// the maximum amount of time for an individual RPC never exceeds the
+    /// remaining time of the test
+    auto now = ss::lowres_clock::now();
+    while (now < fiber_state.test_end() && !_cancelled) {
+        if (unlikely(_as.abort_requested())) {
+            throw netcheck_aborted_exception();
+        }
+        /// Measure the time it takes to complete a full round trip
+        co_await m.measure([this, peer, &fiber_state]() mutable {
+            netcheck_request r{
+              .source = _self, .buf = create_sample_blob(_opts.request_size)};
+            return _connections.local()
+              .with_node_client<self_test_rpc_client_protocol>(
+                _self,
+                ss::this_shard_id(),
+                peer,
+                10s,
```

Comment:
I wonder if we should set the timeout to something lower, if we want to quickly detect that we've overwhelmed a peer.  Or, if we would like to permit a big queue of RPCs, then maybe the timeout should be the test duration.